File: bonsai_scxml_engine/src/test/java/de/unibi/citec/clf/bonsai/engine/scxml/StatemachineTest.java
package de.unibi.citec.clf.bonsai.engine.scxml;

import de.unibi.citec.clf.bonsai.engine.LoadingResults;
import de.unibi.citec.clf.bonsai.test.TestListener;
import org.apache.log4j.PropertyConfigurator;
import org.junit.Before;
import org.junit.Test;

import javax.xml.transform.TransformerException;
import java.util.concurrent.TimeoutException;

import static junit.framework.TestCase.assertFalse;
import static junit.framework.TestCase.assertTrue;
import static org.junit.Assert.assertEquals;

public class StatemachineTest {

    private final String PATH_TO_LOGGING_PROPERTIES = getClass().getResource("/testLogging.properties").getPath();

    @Before
    public void initialize() {
        PropertyConfigurator.configure(PATH_TO_LOGGING_PROPERTIES);
    }

    @Test
    public void testVarsInDatamodel() throws TransformerException, TimeoutException {
        final String sm = "variablesInDatamodel.xml";
        final String conf = "TestConfig.xml";

        TestListener test = TestListener.newEndFatal();

        assertTrue(TestTools.testStatemachine(conf, sm, test));
    }

    @Test
    public void testVarsInDatamodelFailure() throws TransformerException, TimeoutException {
        final String sm = "variablesInDatamodelFailure.xml";
        final String conf = "TestConfig.xml";

        TestListener test = TestListener.newEndFatal();

        assertFalse(TestTools.testStatemachine(conf, sm, test));
    }

    @Test
    public void missingSendEventTransition() throws TransformerException, TimeoutException {
        final String sm = "missingSendEventTransition.xml";
        final String conf = "TestConfig.xml";

        LoadingResults res = TestTools.loadStatemachine(sm);
        assertFalse(res.success());
    }

    @Test
    public void hangupTest() throws TransformerException, TimeoutException {
        final String sm = "parallelHangupExample.xml";
        final String conf = "TestConfig.xml";

        TestListener test = TestListener.newEndFatal();
        assertTrue(TestTools.testStatemachine(conf, sm, test));
    }

    @Test
    public void sendEventTransition() throws TransformerException, TimeoutException {
        final String sm = "sendEventTransition.xml";
        final String conf = "TestConfig.xml";

        LoadingResults res = TestTools.loadStatemachine(sm);
        assertTrue(res.success());
    }

    @Test
    public void slotMapping() throws TransformerException, TimeoutException {
        final String sm = "slotMapping.xml";
        final String conf = "TestConfig.xml";

        TestListener test = TestListener.newEndFatal();
        assertTrue(TestTools.testStatemachine(conf, sm, test));
    }

    @Test
    public void slotMappingRegex() throws TransformerException, TimeoutException {
        final String sm = "slotMappingRegex.xml";
        final String conf = "TestConfig.xml";

        TestListener test = TestListener.newEndFatal();
        assertTrue(TestTools.testStatemachine(conf, sm, test));
    }

    @Test
    public void slotMappingRegexNoValidate() throws TransformerException, TimeoutException {
        final String sm = "slotMappingRegexNoValidate.xml";
        final String conf = "TestConfig.xml";

        TestListener test = TestListener.newEndFatal();
        //failing for now 
//        assertTrue(TestTools.testStatemachine(conf, sm, test));
    }

    @Test
    public void slotMappingnoValidate() throws TransformerException, TimeoutException {
        final String sm = "slotMappingNoValidate.xml";
        final String conf = "TestConfig.xml";

        TestListener test = TestListener.newEndFatal();
        assertTrue(TestTools.testStatemachine(conf, sm, test));
    }

    @Test
    public void testMissingTransition() throws TransformerException, TimeoutException {
        final String sm = "missingTransition.xml";

        LoadingResults res = TestTools.loadStatemachine(sm);

        assertFalse(res.success());

        assertEquals(0,res.configurationResults.getConfigurationExceptions().size());
        assertEquals(0,res.configurationResults.otherExceptions.size());
        assertEquals(1,res.validationResult.transitionNotFoundException.size());
        assertEquals(0,res.stateMachineResults.numErrors());
        assertEquals(0,res.stateMachineResults.numWarnings());
        assertEquals(0,res.loadingExceptions.size());
    }

    @Test
    public void testMissingIncomingTransition() throws TransformerException, TimeoutException {
        final String sm = "missingIncomingTransition.xml";

        LoadingResults res = TestTools.loadStatemachine(sm);

        assertFalse(res.success());

        assertEquals(0,res.configurationResults.getConfigurationExceptions().size());
        assertEquals(0,res.configurationResults.otherExceptions.size());
        assertEquals(0,res.validationResult.transitionNotFoundException.size());
        assertEquals(1,res.validationResult.unreachedStates.size());
        assertEquals(0,res.stateMachineResults.numErrors());
        assertEquals(0,res.stateMachineResults.numWarnings());
        assertEquals(0,res.loadingExceptions.size());
    }

    @Test
    public void testMissingParameter() throws TransformerException, TimeoutException {
        final String sm = "missingParameter.xml";

        LoadingResults res = TestTools.loadStatemachine(sm);

        assertFalse(res.success());

        assertEquals(0,res.configurationResults.getConfigurationExceptions().size());
        assertEquals(0,res.configurationResults.otherExceptions.size());
        assertEquals(0,res.validationResult.transitionNotFoundException.size());
        assertEquals(1,res.stateMachineResults.numErrors());
        // general "Error configuring skill" is a warning :C
        assertEquals(1,res.stateMachineResults.numWarnings());
        assertEquals(0,res.loadingExceptions.size());
    }

    @Test
    public void testOptionalParameter() throws TransformerException, TimeoutException {
        final String sm = "optionalParameter.xml";
        final String conf = "TestConfig.xml";

        LoadingResults res = TestTools.loadStatemachine(sm);
        assertTrue(res.success());

        TestListener test = TestListener.newEndFatal();
        assertTrue(TestTools.testStatemachine(conf, sm, test));
    }

}

File: bonsai_scxml_engine/src/test/java/de/unibi/citec/clf/bonsai/engine/scxml/StatemachineTest.kt
package de.unibi.citec.clf.bonsai.engine.scxml

import de.unibi.citec.clf.bonsai.test.TestListener.Companion.newEndFatal
import de.unibi.citec.clf.bonsai.test.TestListener.Companion.newEndFatalSkillCounter
import junit.framework.TestCase
import org.apache.log4j.PropertyConfigurator
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import java.util.concurrent.TimeoutException
import javax.xml.transform.TransformerException

class StatemachineTest {
    private val PATH_TO_LOGGING_PROPERTIES = javaClass.getResource("/testLogging.properties").path
    @Before
    fun initialize() {
        PropertyConfigurator.configure(PATH_TO_LOGGING_PROPERTIES)
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun reusedVariables() {
        var sm = "variablesReused.xml"
        var conf = "TestConfig.xml"
        var test = newEndFatalSkillCounter(mutableMapOf(
                "dialog.Talk" to 4)
        )

        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun moreVariables() {
        var sm = "variables.xml"
        var conf = "TestConfig.xml"
        var test = newEndFatalSkillCounter(mutableMapOf(
                "dialog.Talk" to 2)
        )
        // Check if test fails
        TestCase.assertFalse(TestTools.testStatemachine(conf, sm, test))

        test = newEndFatalSkillCounter(mutableMapOf(
                "dialog.Talk" to 3)
        )
        // Check if spoken 3 times
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))
        sm = "variables2.xml"
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))

        // Check End
        sm = "variables3.xml"
        test = newEndFatal()
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))

    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun testVarsInDatamodel() {
        val sm = "variablesInDatamodel.xml"
        val conf = "TestConfig.xml"
        val test = newEndFatal()
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun testVarsInDatamodelFailure() {
        val sm = "variablesInDatamodelFailure.xml"
        val conf = "TestConfig.xml"
        val test = newEndFatal()
        TestCase.assertFalse(TestTools.testStatemachine(conf, sm, test))
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun missingSendEventTransition() {
        val sm = "missingSendEventTransition.xml"
        val conf = "TestConfig.xml"
        val res = TestTools.loadStatemachine(sm)
        TestCase.assertFalse(res.success())
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun hangupTest() {
        val sm = "parallelHangupExample.xml"
        val conf = "TestConfig.xml"
        val test = newEndFatal()
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun includeSendEventTransition() {
        val sm = "includeSend.xml"
        val conf = "TestConfig.xml"
        val res = TestTools.loadStatemachine(sm)
        TestCase.assertTrue(res.success())
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun sendEventTransition() {
        val sm = "sendEventTransition.xml"
        val conf = "TestConfig.xml"
        val res = TestTools.loadStatemachine(sm)
        TestCase.assertTrue(res.success())
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun slotMapping() {
        val sm = "slotMapping.xml"
        val conf = "TestConfig.xml"
        val test = newEndFatal()
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun slotMappingRegex() {
        val sm = "slotMappingRegex.xml"
        val conf = "TestConfig.xml"
        val test = newEndFatal()
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun slotMappingRegexNoValidate() {
        val sm = "slotMappingRegexNoValidate.xml"
        val conf = "TestConfig.xml"
        val test = newEndFatal()
        //failing for now
//        assertTrue(TestTools.testStatemachine(conf, sm, test));
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun slotMappingnoValidate() {
        val sm = "slotMappingNoValidate.xml"
        val conf = "TestConfig.xml"
        val test = newEndFatal()
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun testMissingTransition() {
        val sm = "missingTransition.xml"
        val res = TestTools.loadStatemachine(sm)
        TestCase.assertFalse(res.success())
        Assert.assertEquals(0, res.configurationResults.configurationExceptions.size.toLong())
        Assert.assertEquals(0, res.configurationResults.otherExceptions.size.toLong())
        Assert.assertEquals(1, res.validationResult.transitionNotFoundException.size.toLong())
        Assert.assertEquals(0, res.stateMachineResults.numErrors().toLong())
        Assert.assertEquals(0, res.stateMachineResults.numWarnings().toLong())
        Assert.assertEquals(0, res.loadingExceptions.size.toLong())
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun testMissingParameter() {
        val sm = "missingParameter.xml"
        val res = TestTools.loadStatemachine(sm)
        TestCase.assertFalse(res.success())
        Assert.assertEquals(0, res.configurationResults.configurationExceptions.size.toLong())
        Assert.assertEquals(0, res.configurationResults.otherExceptions.size.toLong())
        Assert.assertEquals(0, res.validationResult.transitionNotFoundException.size.toLong())
        Assert.assertEquals(1, res.stateMachineResults.numErrors().toLong())
        // general "Error configuring skill" is a warning :C
        Assert.assertEquals(1, res.stateMachineResults.numWarnings().toLong())
        Assert.assertEquals(0, res.loadingExceptions.size.toLong())
    }

    @Test
    @Throws(TransformerException::class, TimeoutException::class)
    fun testOptionalParameter() {
        val sm = "optionalParameter.xml"
        val conf = "TestConfig.xml"
        val res = TestTools.loadStatemachine(sm)
        TestCase.assertTrue(res.success())
        val test = newEndFatal()
        TestCase.assertTrue(TestTools.testStatemachine(conf, sm, test))
    } //    @Test
    //    public void testParallelWorking() throws TransformerException, TimeoutException {
    //        final String sm = "parallelWorking.xml";
    //        final String conf = "TestConfig.xml";
    //
    //        LoadingResults res = TestTools.loadStatemachine(sm);
    //        assertTrue(res.success());
    //
    //        TestListener test = TestListener.newEndFatal();
    //        assertTrue(TestTools.testStatemachine(conf, sm, test));
    //    }
    //
    //    @Test
    //    public void testParallelBugged() throws TransformerException, TimeoutException {
    //        final String sm = "parallelBugged.xml";
    //        final String conf = "TestConfig.xml";
    //
    //        LoadingResults res = TestTools.loadStatemachine(sm);
    //        assertTrue(res.success());
    //
    //        TestListener test = TestListener.newEndFatal();
    //        assertTrue(TestTools.testStatemachine(conf, sm, test));
    //    }
}
